lib/control/control-main.c

```diff
@@ -26,18 +26,61 @@
 #include "control-server.h"
 #include "control-commands.h"
 
-static ControlServer *control_server;
+#include <iv.h>
+#include <iv_event.h>
+
+typedef struct _ControlServerThread
+{
+  GThread *thread;
+  MainLoop *main_loop;
+  const gchar *control_name;
+  struct iv_event stop_requested;
+} ControlServerThread;
+
+static ControlServerThread control_server_thread;
+
+static void
+_thread_stop()
+{
+  iv_quit();
+}
+
+static void
+_register_stop_requested_event(ControlServerThread *self)
+{
+  IV_EVENT_INIT(&self->stop_requested);
+  self->stop_requested.handler = _thread_stop;
+  iv_event_register(&self->stop_requested);
+}
+
+static void
+_thread(gpointer user_data)
+{
+  ControlServerThread *cs_thread = (ControlServerThread *)user_data;
+  ControlServer *control_server = control_server_new(cs_thread->control_name,
+                                                     control_register_default_commands(cs_thread->main_loop));
+  iv_init();
+  {
+    _register_stop_requested_event(cs_thread);
+    control_server_start(control_server);
+    iv_main();
```

Comment:
I would probably extract these init init/deinit functions, rather than inlineing them here.

Reply:
This implementation has a pro over the extracted init/deinit: no need to store the control_server, it is a local variable. I think it is as readable as the other, but I'll factor out init/deinit functions.